keys.go

```diff
@@ -3,548 +3,196 @@ package publiccode
-	"github.com/rivo/uniseg"
 	"github.com/thoas/go-funk"
 )
 
-func (p *Parser) decodeBool(key string, boolValue bool) (err error) {
-	switch key {
-	case "localisation/localisationReady":
-		p.PublicCode.Localisation.LocalisationReady = boolValue
-	case "it/conforme/lineeGuidaDesign":
-		p.PublicCode.It.Conforme.LineeGuidaDesign = boolValue
-	case "it/conforme/modelloInteroperabilita":
-		p.PublicCode.It.Conforme.ModelloInteroperabilita = boolValue
-	case "it/conforme/misureMinimeSicurezza":
-		p.PublicCode.It.Conforme.MisureMinimeSicurezza = boolValue
-	case "it/conforme/gdpr":
-		p.PublicCode.It.Conforme.GDPR = boolValue
-	case "it/piattaforme/spid":
-		p.PublicCode.It.Piattaforme.Spid = boolValue
-	case "it/piattaforme/pagopa":
-		p.PublicCode.It.Piattaforme.Pagopa = boolValue
-	case "it/piattaforme/cie":
-		p.PublicCode.It.Piattaforme.Cie = boolValue
-	case "it/piattaforme/anpr":
-		p.PublicCode.It.Piattaforme.Anpr = boolValue
+// validateFields validates publiccode with additional rules not validatable
+// with a simple YAML schema.
+// It returns any error encountered as ValidationErrors.
+func (p *Parser) validateFields() error {
+	var ve ValidationErrors
+	var err error
 
-	default:
-		return newErrorInvalidValue(key, "Unexpected boolean key")
+	if (p.PublicCode.URL != nil) {
+		if reachable, err := p.isReachable((*url.URL)(p.PublicCode.URL)); !reachable {
+			ve = append(ve, newValidationError("url", "'%s' not reachable: %s", p.PublicCode.URL, err.Error()))
+		}
+		if !vcsurl.IsRepo((*url.URL)(p.PublicCode.URL)) {
+			ve = append(ve, newValidationError("url", "is not a valid code repository"))
+		}
 	}
-	return
-}
 
-func (p *Parser) decodeString(key string, value string) (err error) {
-	switch {
-	case key == "publiccodeYmlVersion":
-		// strip legacy URI prefix
-		value = strings.Replace(value, "http://w3id.org/publiccode/version/", "", 1)
-
-		if !funk.Contains(SupportedVersions, value) {
-			return newErrorInvalidValue(key, "version %s not supported", value)
-		}
-		p.PublicCode.PubliccodeYamlVersion = value
-	case key == "name":
-		p.PublicCode.Name = value
-	case key == "applicationSuite":
-		p.PublicCode.ApplicationSuite = value
-	case key == "url":
-		// Check that the supplied URL is valid and exists.
-		p.PublicCode.URLString, p.PublicCode.URL, err = p.checkURL(key, value)
+	// Check that the supplied URL matches the source repository, if known.
+	if p.RemoteBaseURL != "" {
+		url1, err := url.Parse(p.RemoteBaseURL)
 		if err != nil {
-			return err
+			ve = append(ve, newValidationError("", err.Error()))
 		}
-
-		// Check that the supplied URL points to a repository.
-		if !vcsurl.IsRepo(p.PublicCode.URL) {
-			return newErrorInvalidValue(key, "invalid repository URL: %s", value)
+		repo1 := vcsurl.GetRepo(url1)
+		repo2 := vcsurl.GetRepo((*url.URL)(p.PublicCode.URL))
+		if repo1 == nil {
+			ve = append(ve, newValidationError(
+				"", "failed to detect repo for remote-base-url: %s\n", url1.String()),
```

Comment:
perhaps I miss something but we should always have key field valued. In this file there are others occurrences.

Reply:
This particular error doesn't have a key if refers to because remote-base-url gets passed out of band and it's not in the YAML file (as an argument in case of pcvalidate and as a struct field in case of the library). 

In general, there can be cases where key is empty, when the error is not about a specific key, fe. when the encoding is not correct.